File: examples/phoenix/lib/d3_ex_demo_web/components/charts/stream_chart.ex
defmodule D3ExDemoWeb.Components.Charts.StreamChart do
  @moduledoc """
  Example Phoenix `stream/3` ↔ D3 bridge built on `D3Ex.Component`. Lives in the
  demo app, not the library — D3Ex itself ships only the bridge primitives.

  Pairs with `assets/js/hooks/stream_chart.js`. The hook uses a MutationObserver
  on the hidden `phx-update="stream"` feed and re-feeds D3 whenever Phoenix
  mutates the list.

  Why this pattern is nice:

  - `stream_insert(socket, :points, p, limit: -200)` caps memory server-side —
    no hand-rolled ring buffer in assigns.
  - Reconnect semantics fall out of LiveView's stream protocol for free.
  - Same `stream_insert`/`stream_delete` you already use for tables.

  ## Example

      def mount(_params, _session, socket) do
        if connected?(socket), do: :timer.send_interval(250, self(), :tick)

        socket =
          socket
          |> stream_configure(:points, dom_id: &"pt-\#{&1.t}")
          |> stream(:points, [])

        {:ok, socket}
      end

      def handle_info(:tick, socket) do
        t = System.system_time(:millisecond)
        point = %{t: t, y: :math.sin(t / 1000)}
        {:noreply, stream_insert(socket, :points, point, limit: -200)}
      end

      def render(assigns) do
        ~H\"\"\"
        <.stream_chart id="ticks" stream={@streams.points} x_key={:t} y_key={:y} />
        \"\"\"
      end

  ## Options

  - `stream` - A `Phoenix.LiveView.LiveStream` reference (required)
  - `x_key`, `y_key` - Atom keys into each item (default: `:x`, `:y`)
  - `series_key` - Optional key for multi-line grouping
  - Plus the standard `width`/`height`/`margin`/`color_scheme`/`curve_type`/
    `show_points`/`show_area`/`show_grid`/`animation_duration`/`point_radius`

  ## Limitations

  Numeric x/y only — `data-*` attributes are strings, coerced via `Number(...)`.
  Use millisecond integers for timestamps (`System.system_time(:millisecond)`).
  """

  use D3Ex.Component

  @impl true
  def default_config do
    %{
      width: 800,
      height: 400,
      margin: %{top: 20, right: 80, bottom: 40, left: 60},
      color_scheme: "schemeCategory10",
      curve_type: "monotone",
      show_points: true,
      show_area: false,
      show_grid: true,
      animation_duration: 750,
      point_radius: 4
    }
  end

  @impl true
  def prepare_assigns(assigns) do
    unless Map.has_key?(assigns, :stream) do
      raise ArgumentError, """
      StreamChart requires a `:stream` prop pointing at a
      `Phoenix.LiveView.LiveStream`. See `Phoenix.LiveView.stream/3`.

          <.stream_chart id="ticks" stream={@streams.points} x_key={:t} y_key={:y} />
      """
    end

    assigns
    |> Map.put_new(:x_key, :x)
    |> Map.put_new(:y_key, :y)
    |> Map.put_new(:series_key, nil)
    |> Map.put_new(:x_label, nil)
    |> Map.put_new(:y_label, nil)
  end

  @impl true
  def render(assigns) do
    ~H"""
    <div
      id={@id}
      phx-hook="D3Stream"
      data-config={encode_config(Map.merge(@config, %{
        x_key: @x_key,
        y_key: @y_key,
        series_key: @series_key,
        x_label: @x_label,
        y_label: @y_label
      }))}
      class="d3-stream-chart"
      style={"width: #{@config.width}px; height: #{@config.height}px;"}
    >
      <div id={"#{@id}-svg"} phx-update="ignore">
        <svg width={@config.width} height={@config.height}></svg>
      </div>

      <div
        id={"#{@id}-feed"}
        phx-update="stream"
        data-stream-feed
        style="display:none"
      >
        <div
          :for={{dom_id, item} <- @stream}
          id={dom_id}
          data-stream-item
          data-x={Map.get(item, @x_key)}
          data-y={Map.get(item, @y_key)}
          data-series={@series_key && Map.get(item, @series_key)}
        />
      </div>
    </div>
    """
  end
end

File: examples/phoenix/assets/js/hooks/stream_chart.js
/**
 * D3 Stream hook — example component for the demo app.
 *
 * Bridges Phoenix `stream/3` to a D3 line chart via `createStreamD3Hook`,
 * which handles the MutationObserver wiring on `[data-stream-feed]`. The
 * Elixir component renders a hidden `<div phx-update="stream">` feed of
 * `<div data-stream-item>` nodes; this hook just defines what to do when
 * `this.data` changes.
 *
 * Server-side memory is bounded by `stream_insert(socket, :points, p, limit:
 * N)`; reconnect semantics come from Phoenix's stream protocol.
 *
 * Numeric x/y only — the default `parseRow` coerces `data-x`/`data-y` via
 * `Number()`. For non-numeric data, pass a custom `parseRow` to
 * `createStreamD3Hook`.
 *
 * Pairs with `D3ExDemoWeb.Components.Charts.StreamChart`.
 */

import { createStreamD3Hook } from "../../../../../priv/static/js/d3_hooks.js";

export const D3Stream = {
  ...createStreamD3Hook({
    onMount()  { this.initChart() },
    onUpdate() { this.renderChart() },
  }),

  initChart() {
    const d3 = window.d3;
    const { width, height, margin } = this.config;

    this.svg = d3.select(this.el).select('svg');
    this.g = this.svg.append('g')
      .attr('transform', `translate(${margin.left},${margin.top})`);

    this.innerWidth = width - margin.left - margin.right;
    this.innerHeight = height - margin.top - margin.bottom;

    this.xScale = d3.scaleLinear().range([0, this.innerWidth]);
    this.yScale = d3.scaleLinear().range([this.innerHeight, 0]);
    this.colorScale = d3.scaleOrdinal(d3[this.config.color_scheme] || d3.schemeCategory10);

    this.xAxis = this.g.append('g')
      .attr('class', 'x-axis')
      .attr('transform', `translate(0,${this.innerHeight})`);

    this.yAxis = this.g.append('g').attr('class', 'y-axis');

    if (this.config.show_grid) {
      this.xGrid = this.g.append('g')
        .attr('class', 'grid x-grid')
        .style('stroke', '#e0e0e0')
        .style('stroke-opacity', 0.3);

      this.yGrid = this.g.append('g')
        .attr('class', 'grid y-grid')
        .style('stroke', '#e0e0e0')
        .style('stroke-opacity', 0.3);
    }

    this.linesGroup = this.g.append('g').attr('class', 'lines');
    this.pointsGroup = this.g.append('g').attr('class', 'points');

    this.renderChart();
  },

  renderChart() {
    const d3 = window.d3;
    const { x_key, y_key, series_key, curve_type, show_points, show_area, animation_duration } = this.config;

    if (this.data.length === 0) {
      this.linesGroup.selectAll('.line').remove();
      this.linesGroup.selectAll('.area').remove();
      this.pointsGroup.selectAll('circle').remove();
      return;
    }

    let seriesData;
    if (series_key) {
      const grouped = d3.group(this.data, d => d[series_key]);
      seriesData = Array.from(grouped, ([key, values]) => ({
        key,
        values: values.sort((a, b) => a[x_key] - b[x_key]),
      }));
    } else {
      seriesData = [{
        key: 'default',
        values: this.data.slice().sort((a, b) => a[x_key] - b[x_key]),
      }];
    }

    this.xScale.domain(d3.extent(this.data, d => d[x_key]));
    this.yScale.domain([
      d3.min(this.data, d => d[y_key]),
      d3.max(this.data, d => d[y_key]),
    ]);

    this.xAxis.transition().duration(animation_duration).call(d3.axisBottom(this.xScale));
    this.yAxis.transition().duration(animation_duration).call(d3.axisLeft(this.yScale));

    if (this.config.show_grid) {
      this.xGrid.transition().duration(animation_duration)
        .call(d3.axisBottom(this.xScale).tickSize(-this.innerHeight).tickFormat(''));

      this.yGrid.transition().duration(animation_duration)
        .call(d3.axisLeft(this.yScale).tickSize(-this.innerWidth).tickFormat(''));
    }

    const curveTypes = {
      'linear': d3.curveLinear,
      'monotone': d3.curveMonotoneX,
      'step': d3.curveStep,
    };

    const line = d3.line()
      .x(d => this.xScale(d[x_key]))
      .y(d => this.yScale(d[y_key]))
      .curve(curveTypes[curve_type] || d3.curveMonotoneX);

    let area;
    if (show_area) {
      area = d3.area()
        .x(d => this.xScale(d[x_key]))
        .y0(this.innerHeight)
        .y1(d => this.yScale(d[y_key]))
        .curve(curveTypes[curve_type] || d3.curveMonotoneX);
    }

    // Paths snap to the new shape instantly — D3's default `d` interpolator is
    // string-based and produces visual chaos when the vertex count changes
    // (which is every tick in a streaming chart). The smooth motion the eye
    // wants comes from the axis transition above + a short tick interval.
    if (show_area) {
      const areas = this.linesGroup.selectAll('.area').data(seriesData, d => d.key);
      areas.exit().remove();
      areas.enter()
        .append('path')
        .attr('class', 'area')
        .attr('fill-opacity', 0.2)
        .merge(areas)
        .attr('d', d => area(d.values))
        .attr('fill', d => this.colorScale(d.key));
    }

    const lines = this.linesGroup.selectAll('.line').data(seriesData, d => d.key);
    lines.exit().remove();

    lines.enter()
      .append('path')
      .attr('class', 'line')
      .attr('fill', 'none')
      .attr('stroke-width', 2)
      .merge(lines)
      .attr('d', d => line(d.values))
      .attr('stroke', d => this.colorScale(d.key));

    if (show_points) {
      const points = this.pointsGroup
        .selectAll('circle')
        .data(this.data, d => `${d[x_key]}-${d[y_key]}`);

      points.exit().remove();

      // Place new points at their correct (cx, cy) on enter — otherwise they
      // appear at the SVG origin and animate across the chart.
      points.enter()
        .append('circle')
        .attr('r', this.config.point_radius)
        .attr('stroke', '#fff')
        .attr('stroke-width', 1.5)
        .attr('cx', d => this.xScale(d[x_key]))
        .attr('cy', d => this.yScale(d[y_key]))
        .attr('fill', d => series_key ? this.colorScale(d[series_key]) : this.colorScale('default'))
        .merge(points)
        .attr('cx', d => this.xScale(d[x_key]))
        .attr('cy', d => this.yScale(d[y_key]))
        .attr('fill', d => series_key ? this.colorScale(d[series_key]) : this.colorScale('default'));
    }
  },
};
